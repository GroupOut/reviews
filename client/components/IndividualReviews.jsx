import React from 'react';

class IndividualReviews extends React.Component {
  constructor(props) {
    super(props)
    
    this.sortReviewsByDate = this.sortReviewsByDate.bind(this);
    this.formatDateToBeSorted = this.formatDateToBeSorted.bind(this);
    this.formatUsername = this.formatUsername.bind(this);
    this.state = {
      reviews: {deals_id : 12, helpful_score : 0, id : 7, name : "Full-Day Waterfall Rappelling for One, Two, or Four at North Ridge Mountain Guides", relevancy_score : 0, review_date : "2017-07-09T22:36:49.000Z", review_score : 4, review_text : "Something to start", reviews_count : 17, top_reviewer : 1, username : "Lanie Igonet"},
      visableReviews: [{deals_id : 12, helpful_score : 0, id : 7, name : "Full-Day Waterfall Rappelling for One, Two, or Four at North Ridge Mountain Guides", relevancy_score : 0, review_date : "2017-07-09T22:36:49.000Z", review_score : 4, review_text : "Something to start", reviews_count : 17, top_reviewer : 1, username : "Lanie Igonet"}],
      // reviewOne: {deals_id : 12, helpful_score : 0, id : 7, name : "Full-Day Waterfall Rappelling for One, Two, or Four at North Ridge Mountain Guides", relevancy_score : 0, review_date : "2017-07-09T22:36:49.000Z", review_score : 4, review_text : "Something to start", reviews_count : 17, top_reviewer : 1, username : "Lanie Igonet"},
      // reviewTwo: {deals_id : 12, helpful_score : 0, id : 7, name : "Full-Day Waterfall Rappelling for One, Two, or Four at North Ridge Mountain Guides", relevancy_score : 0, review_date : "2017-07-09T22:36:49.000Z", review_score : 4, review_text : "Something to start", reviews_count : 17, top_reviewer : 1, username : "Lanie Igonet"},
      // reviewThree: {deals_id : 12, helpful_score : 0, id : 7, name : "Full-Day Waterfall Rappelling for One, Two, or Four at North Ridge Mountain Guides", relevancy_score : 0, review_date : "2017-07-09T22:36:49.000Z", review_score : 4, review_text : "Something to start", reviews_count : 17, top_reviewer : 1, username : "Lanie Igonet"},
    }
  }

  componentDidMount() {
    console.log('COMPONENT STATE', this.state)
  }

  componentWillReceiveProps(nextProps){
    if(nextProps.reviews !== this.props.reviews){
        this.sortReviewsByDate(nextProps.reviews);
        var visableReviews = [];
        nextProps.reviews.forEach((review, index) => {
          if(index < 3) {
            visableReviews.push(review);
          }
        });
        this.setState({
          reviews: nextProps.reviews,
          visableReviews: visableReviews,
          // reviewOne: nextProps.reviews[0],
          // reviewTwo: nextProps.reviews[1],
          // reviewThree: nextProps.reviews[2],
        });
    }
  }

  formatDateToBeSorted(date) {
    //string
    var arr = [];
    var year = date.slice(0, 4);
    var month = date.slice(5, 7);
    var day = date.slice(8, 10);
    var hour = date.slice(11, 13);
    var minute = date.slice(14, 16);
    var second = date.slice(17, 19);
    arr.push(year, month, day, hour, minute, second,);
    var returnDate = Number(arr.join(''));
    return returnDate;
  }

  sortReviewsByDate(reviewsArr) {
    var that = this;
    var beforeSort = reviewsArr;
    console.log('BEFORE SORT', beforeSort);
    var afterSort = reviewsArr.sort(function (a, b) {
      let formattedA = that.formatDateToBeSorted(a.review_date);
      a.sortBy = formattedA
      let formattedB = that.formatDateToBeSorted(b.review_date);
      b.sortBy = formattedB
      return b.sortBy - a.sortBy
    });
    afterSort.forEach(item => {
      delete item.sortBy;
    });
  }


  formatUsername(user) {
    let index = user.indexOf(' ');
    user = user.slice(0, index + 2)
    user = user.concat('.');
    return user;
  }

  render () {
    return (
      <div>
        <h3>Relevant Reviews</h3>

        {this.state.visableReviews.map(review => 
          <div className='firstReview'>
            <div className='firstHeader'>
              <p>{this.formatUsername(review.username)} {review.reviews_count} reviews</p>
              <p>{review.review_score} SCORE DATE {review.review_date} </p>
            </div>
            <div className='firstBody'>
              <p>{review.review_text}</p>
              <p> helpfulButton</p>
            </div>
          </div>

        )}
      </div>
    )
  }
  
}


export default IndividualReviews;