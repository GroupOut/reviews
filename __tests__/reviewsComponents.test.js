const sum = require('../sum');
import Reviews from '../client/components/reviews';
import IndividualReviews from '../client/components/IndividualReviews';
import React from 'react';
import ReactDOM from 'react-dom';
import render from "react-test-renderer";
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import styled from 'styled-components';
import sinon from 'sinon';

configure({ adapter: new Adapter() });

test('adds 1 + 2 to equal 3', function () {
  expect(sum(1, 2)).toBe(3);
})

// test('check that formatDateToBeSorted correctly formats dates to a number', function() {
//   expect(IndividualReviews.formatDateToBeSorted('2018-01-31 22:18:01')).toBe(20180131221801);
// })

describe('Reviews', () => {
  it('the ReviewsHeaderTitle element should exist', () => {
    const wrapper = shallow(<Reviews />);
    expect(wrapper.find('ReviewsHeaderTitle').exists()).toEqual(true);
  });
  
  it('Snapshot of header', () => {
    const wrapper = shallow(<Reviews />);
    expect(wrapper.find('.header')).toMatchSnapshot();
  });
  // it('snapshot works correctly', () => {
  //   const wrapper = shallow(<Reviews />);
  //   expect(wrapper).toMatchSnapshot();
  // })
  // it('the ReviewsHeaderTitle should contain the text Customer Reviews', () => {
  //   const wrapper = shallow(<Reviews />);
  //   expect(wrapper.find('ReviewsHeaderTitle').text()).toEqual('Customer Reviews')
  // });
  it('should call componentWillMount', () => {
    sinon.spy(Reviews.prototype, 'componentWillMount');
    const wrapper = shallow(<Reviews />);
    expect(Reviews.prototype.componentWillMount.calledOnce).toEqual(true);
  })
  // it('shoud call handleSubmit when the handleSubmit button is called', () => {
  //   sinon.spy(Reviews.prototype, 'handleSubmit');
  //   const wrapper = shallow(<Reviews />);
  //   wrapper.find('form').simulate('submit', {e: 9});
  //   expect(Reviews.prototype.handleSubmit.calledOnce).toEqual(true);
  // })
});

describe('IndividualReviews', () => {
  // it('snapshot works correctly', () => {
  //   const wrapper = shallow(<IndividualReviews />);
  //   expect(wrapper).toMatchSnapshot();
  // })
  it('the ReviewCount element should exist', () => {
    const wrapper = shallow(<IndividualReviews />);
    expect(wrapper.find('ReviewCount').exists()).toEqual(true);
  })
  // it('should alert feature not built yet when see all reviews component is clicked', () => {
  //   window.alert = jest.fn(() => true);
  //   const wrapper = shallow(<IndividualReviews />);
  //   wrapper.find('AllReviews').simulate('click');
  //   expect(window.alert).toHaveBeenCalledWith('feature not built yet');
  // })
  it('should call componentDidMount', () => {
    sinon.spy(IndividualReviews.prototype, 'componentDidMount');
    const wrapper = shallow(<IndividualReviews />);
    expect(IndividualReviews.prototype.componentDidMount.calledOnce).toEqual(true);
  })
  it('shoud call handleHelpfulClicks when the HelpfulButton button is called', () => {
    sinon.spy(IndividualReviews.prototype, 'handleHelpfulClick');
    const wrapper = shallow(<IndividualReviews />);
    wrapper.find('HelpfulButton').simulate('click', {review: 20});
    expect(IndividualReviews.prototype.handleHelpfulClick.calledOnce).toEqual(true);
  })
  it('should call formatUsername on render', () => {
    sinon.spy(IndividualReviews.prototype, 'formatUsername');
    const wrapper = shallow(<IndividualReviews />);

    expect(wrapper.find('ReviewCount').exists()).toEqual(true);
  })

})


// export default class Welcome extends React.Component {
//   render() {fffd
//     return (
//       <div>Hello world</div>
//     );
//   }
// }

// .descrip
