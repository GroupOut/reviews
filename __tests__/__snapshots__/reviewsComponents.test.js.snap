// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndividualReviews snapshot works correctly 1`] = `
<div>
  <ReviewTitle>
    Relevant Reviews
  </ReviewTitle>
  <ReviewWrapper
    className="Review"
  >
    <div
      className="Header"
    >
      <Circle>
        LI
      </Circle>
      <div>
        <Username>
          Lanie I.
        </Username>
        <ReviewCount>
          17
           reviews
        </ReviewCount>
      </div>
      <div>
        <StarRatings
          changeRating={null}
          gradientPathName=""
          ignoreInlineStyles={false}
          isSelectable={false}
          numberOfStars={5}
          rating={4}
          starDimension="14px"
          starEmptyColor="#A5A8AB"
          starHoverColor="rgb(230, 67, 47)"
          starRatedColor="#F7C24A"
          starSpacing="0px"
          svgIconPath="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          svgIconViewBox="0 0 51 48"
          typeOfWidget="Star"
        />
        <ReviewDate>
           
          July 9, 2017
           
        </ReviewDate>
      </div>
    </div>
    <div
      className="Body"
    >
      <ReviewText>
        Something to start
      </ReviewText>
    </div>
    <HelpfulButton
      onClick={[Function]}
    >
      üëç Helpful
    </HelpfulButton>
  </ReviewWrapper>
  <AllReviews
    onClick={[Function]}
  >
    See all reviews
  </AllReviews>
</div>
`;

exports[`Reviews Snapshot of header 1`] = `
<div
  className="header"
>
  <ReviewsHeaderTitle>
    Customer Reviews
  </ReviewsHeaderTitle>
  <h3>
    <StarRatings
      changeRating={null}
      gradientPathName=""
      ignoreInlineStyles={false}
      isSelectable={false}
      numberOfStars={5}
      rating={0}
      starDimension="17px"
      starEmptyColor="#A5A8AB"
      starHoverColor="rgb(230, 67, 47)"
      starRatedColor="#F7C24A"
      starSpacing="0px"
      svgIconPath="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
      svgIconViewBox="0 0 51 48"
      typeOfWidget="Star"
    />
    <NumberOfReviews>
       Ratings
    </NumberOfReviews>
  </h3>
  <BannerWrapper>
    <GreenCircle>
      ‚úî
    </GreenCircle>
    <div>
      <span>
        100% Verified Reviews
      </span>
    </div>
    <div>
      <span>
        All reviews are from people who have redeemed deals with this merchant.
      </span>
    </div>
  </BannerWrapper>
</div>
`;

exports[`Reviews snapshot works correctly 1`] = `
<ReviewsWrapper>
  <div
    className="change reviews"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Change deal:
        <input
          onChange={[Function]}
          type="text"
        />
      </label>
      <input
        className="submitButton"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
  <div
    className="header"
  >
    <ReviewsHeaderTitle>
      Customer Reviews
    </ReviewsHeaderTitle>
    <h3>
      <StarRatings
        changeRating={null}
        gradientPathName=""
        ignoreInlineStyles={false}
        isSelectable={false}
        numberOfStars={5}
        rating={0}
        starDimension="17px"
        starEmptyColor="#A5A8AB"
        starHoverColor="rgb(230, 67, 47)"
        starRatedColor="#F7C24A"
        starSpacing="0px"
        svgIconPath="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
        svgIconViewBox="0 0 51 48"
        typeOfWidget="Star"
      />
      <NumberOfReviews>
         Ratings
      </NumberOfReviews>
    </h3>
    <BannerWrapper>
      <GreenCircle>
        ‚úî
      </GreenCircle>
      <div>
        <span>
          100% Verified Reviews
        </span>
      </div>
      <div>
        <span>
          All reviews are from people who have redeemed deals with this merchant.
        </span>
      </div>
    </BannerWrapper>
  </div>
  <div
    className="Review bodies"
  >
    <IndividualReviews
      reviews={
        Array [
          Object {},
          Object {},
        ]
      }
    />
  </div>
</ReviewsWrapper>
`;
